src/resolution/resolver.ts

```diff
@@ -51,103 +51,128 @@ const _resolveRequest = <T>(requestScope: interfaces.RequestScope) =>
         });
 
     } else {
-
-        let result: undefined | T | Promise<T> | (T | Promise<T>)[];
-
         if (request.target.isOptional() && bindings.length === 0) {
             return undefined;
         }
 
         const binding = bindings[0];
-        const isSingleton = binding.scope === BindingScopeEnum.Singleton;
-        const isRequestSingleton = binding.scope === BindingScopeEnum.Request;
 
-        if (isSingleton && binding.activated) {
-            return binding.cache;
-        }
+        return _resolveBinding<T>(requestScope, request, binding);
+    }
+};
 
-        if (
-            isRequestSingleton &&
-            requestScope !== null &&
-            requestScope.has(binding.id)
-        ) {
-            return requestScope.get(binding.id);
-        }
+const _getResolvedFromBinding = <T>(
+    requestScope: interfaces.RequestScope,
+    request: interfaces.Request,
+    binding:interfaces.Binding<T>) => {
+    let result: T | Promise<T>;
+    const childRequests = request.childRequests;
+    if (binding.type === BindingTypeEnum.ConstantValue) {
+        result = binding.cache!;
+        binding.activated = true;
+    } else if (binding.type === BindingTypeEnum.Function) {
+        result = binding.cache!;
+        binding.activated = true;
+    } else if (binding.type === BindingTypeEnum.Constructor) {
+        result = binding.implementationType as T;
+    } else if (binding.type === BindingTypeEnum.DynamicValue && binding.dynamicValue !== null) {
+        result = invokeFactory(
+            "toDynamicValue",
+            binding.serviceIdentifier,
+            () => (binding.dynamicValue!)(request.parentContext)
```

Comment:
It's easy not to notice about this notation (`!`) I often avoid it in favor of the casting, but it's just my opinion

Reply:
Exclamation marks stand out to me ! 
Feel free to put back !

Reply:
I am refactoring and they should go

Reply:
Gone !